### DISCUSSION TITLE
Fixed Problem using filament sensor on TFT35 in Touch mode

### DISCUSSION THREAD

**Discussion by StonXX3D** [bigtreetech/BIGTREETECH-TouchScreenFirmware#2831] [General]
Hello,
I have two 3D printers with Bigtreetech boards,
one with the Octopus V1.0, the second with the SKR 1.4 Turbo.
A TFT 35 is connected to both.

A Smart Filament Sensor V1 is connected to the Octo and a switch is connected to the Turbo as a sensor.
So far I had the problem that these sensors only worked in Marlin mode and triggered the M600 filament change.
In TFT mode it didn't work on either board.
Now I've found a Marlin option that allows the sensors to also work in touch mode.

In Marlin 2.1.2 there is the option in Advanced.h

#define M114_LEGACY // M114 used to synchronize on every call. Enable if needed.

If this is activated, the filament handling also works in touch mode!

  Greetings

**Comment by burneme** [comment]
Hello,
I faced similar problems with the sensor. Thank you for the hint. However I found, that the actual Prusa slicer always sets the E-value to zero using "G92 E0" whenever it reaches values around 8.4, for a new layer anyhow. I didn't see any possibility to change this behaviour, prusa explains it with better resolution for the extruder since the format is single precision float. I just checked Cura, here I can get code which continously  counts up the extruder length. Parallel I'm debugging the code inside "AddonHardware.c" to understand how data is handled.
In my setup (direct extruder) the smart filament sensor is directly below the spool not moving with the extrude (sfs is quite heavy!) and has at least 20cm distance to the extruder. In my setup the printer pulls out filament when visiting corners and afterward doesn't  pull filament when printing in the middle. So I also increased the max 50mm for FIL_RUNOUT_DISTANCE.

Which slicer do you use?
Did you test the function of the sms, means cutting the filament and it recognizes that?
Greetings

**Reply by burneme** [reply]
Another idea:
Since the slicer output is not always to be in the best shape(setting Extruder to zero within the layer):
**Would it be better to addup the E-ejection on our own inside the touchscreen instead asking the mainboard with a runnung M114 and compare this with the SFS pulses?** 
At the first glance I see only advantages.
Greetings

**Comment by StonXX3D** [comment]
Hello,
I use Cura as a slicer, currently. I'm also still experimenting with Bambu Studio.
Unfortunately, I realized that my success was only short-lived.
Unfortunately, in a recent test, the filament breakage detection on the two TFTs no longer triggered.
The whole thing puzzles me and I was already optimistic that the M114 could have been the solution.
This is all very strange, especially since it works when printing in Marlin mode.
I have the same phenomenon with a BTT UPS24V, which does not raise the print head as desired in the event of a power failure.
Only short mains wipers are bridged by the large capacitors on the circuit board.
If I find the time, I'll experiment again here.

**Reply by burneme** [reply]
Hi,
I just tried it using a modification inside **addonHardware.c** together with Cura Slicer-output.
The idea is very easy: whenever I detect a level change from the SFS (the definition for the pin-state in code is "runout") I set my "Credit" to the vale of "infoSettings.runout_distance", then I need to calc my Credit by
 _Credit = infoSettings.runout_distance - posE + lastPosE;_
When this goes below zero, the routine returns **true** and the foreseen procedure takes place. I tested it and works fine.
However at the moment just with Cura, since prusa (i like that slicer really) always places a "g92 e0".
In one pic you can see the current value of credit right in the middle.

![IMG_4092](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/assets/82312497/6458e8bf-0a7e-4454-a1d5-6e09695ca8ef)
![IMG_4089](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/assets/82312497/65e3a83f-5715-494f-8ef9-4dfe66bf2bc7)
Greetings

**Comment by StonXX3D** [comment]
New trial,
I have now connected the SFS directly to the TFT board BTT 35 V3.0 and deactivated the SFS in Marlin.
The signal from the SFS is routed to ground via a 4.7K resistor over a blue LED, the TFT is connected between the resistor and LED, so that the voltage via the LED is limited to 3.3 volts.
With the LED you can now also observe the signal from the SFS.
I discovered that a signal change on my SFS from H > L > H corresponds to a distance of approx. 4 mm, not 7mm.
After changing the length setting in the TFT config to 5mm, it now apparently works, lack of filament triggers the message and there is no interruption in printing as long as the filament is recognized.
I'll keep watching.....

**Comment by burneme** [comment]
the connection SFS directly to the TFT-pcb is the way it should be.
